crates/loader/src/local/assets.rs

```diff
@@ -130,37 +152,45 @@ fn collect_placement(
     }
 
     let walker = WalkDir::new(&abs);
-    let files = walker
+    let dir_entries = walker
         .into_iter()
-        .filter_map(|de| match de {
-            Err(e) => Some(
-                Err(e).with_context(|| format!("Failed to walk directory under {}", abs.display())),
-            ),
-            Ok(dir_entry) => {
-                if dir_entry.file_type().is_file() {
-                    let match_path = dir_entry.path();
-                    match to_relative(match_path, &abs) {
-                        Ok(relative_to_match_root_dst) => {
-                            let guest_dst = relative_guest_path.join(relative_to_match_root_dst);
-                            Some(FileMount::from_exact(match_path, &guest_dst))
-                        }
-                        Err(e) => {
-                            let err = Err(e).with_context(|| {
-                                format!(
-                                    "Failed to establish relative path for '{}'",
-                                    match_path.display()
-                                )
-                            });
-                            Some(err)
-                        }
-                    }
-                } else {
-                    None
-                }
+        .map(|de| de.with_context(|| format!("Failed to walk directory under {}", abs.display())))
+        .collect::<Result<Vec<_>>>()?;
+    let files = dir_entries
+        .iter()
+        .filter_map(|dir_entry| {
+            if !dir_entry.file_type().is_file() {
+                return None;
             }
+            let match_path = dir_entry.path();
+            let relative_to_abs = to_relative(match_path, &abs);
+            if let Err(e) = relative_to_abs {
+                return Some(Err(e).with_context(|| {
+                    format!(
+                        "Failed to establish relative path for '{}' to '{}'",
+                        match_path.display(),
+                        abs.display(),
+                    )
+                }));
+            }
+            let relative_to_rel = to_relative(match_path, &rel);
+            if let Err(e) = relative_to_rel {
+                return Some(Err(e).with_context(|| {
+                    format!(
+                        "Failed to establish relative path for '{}' to '{}'",
+                        match_path.display(),
+                        rel.as_ref().display(),
+                    )
+                }));
+            }
+            let guest_dst = relative_guest_path.join(relative_to_abs.unwrap());
```

Comment:
Although these unwraps are currently infallible because of the `if let` statements above, I'm anxious they would prevent the compiler from flagging it if the previous code changed in a way that made them fallible.

I wonder if we could reformulate the error checking along the lines of

```
let relative_to_rel_definitely_for_realz = match relative_to_rel {
  Ok(r) => r,
  Err(e) => return Some(...,)
};
```

Reply:
Removed related updates in `collect_placement`.